docs/bespoke.md

```diff
@@ -0,0 +1,121 @@
+the legacy `mongo` extension, or on top of the Doctrine ODM for MongoDB,
+suffering from a total lack of integration with the underlying core Drupal CMS.
+
+Starting with 8.x-2.0, such one-off code can be developed on top of the base
+`mongodb` module: unlike earlier releases, 8.x-2.x uses the PHP-standard
+connection methods and options, without deviation, adding only a thin layer of
+Drupal adaptation on top of the standard `mongodb` [extension] and
+[PHP library].
+
+[extension]: http://php.net/mongodb
+[PHP library]: http://mongodb.github.io/mongo-php-library/
+
+
+## Example
+
+The familiar Drupal alias mechanism for databases is available to provide easy,
+string-referenced access to `Client` and `Database` instances through the
+package-provided `ClientFactory` and `DatabaseFactory` services respectively.
+
+Most such code is likely to be service based, so here is an example of a service
+`bar` in module `foo`, using a custom `foo-database` aliased as `foodb`, to keep
+its storage separate from the main database used by the package modules, and its
+logic independent of other Drupal modules.
+
+
+### Per-environment settings
+
+The site local settings file includes the alias definition, binding it to the
+actual database credentials, allowing for per-environment configuration:
+
+    <?php
+    // settings.local.php
+    $settings['mongodb'] = [
+      'clients' => [
+        // Client alias => constructor parameters.
+        'default' => [
+          'uri' => 'mongodb://localhost:27017',
+          'uriOptions' => [],
+          'driverOptions' => [],
+        ],
+      ],
+      'databases' => [
+        // Collection alias => [ client_alias, collection_name ]
+        'default' => ['default', 'drupal'],
+        'logger' => ['default', 'logger'],
+        'foodb' => ['default', 'foo-database'],
+      ],
+    ];
+
+With such a configuration, the `foodb` alias is available to all MongoDB-using
+modules in the site, possibly pointing to different databases depending on the
+environment (development, staging, production...).
+
+
+### Service-based module adapter
+
+The `foo.services.yml` service file for the bespoke `foo.module` can then
+reference `foodb` to access the database with a constant alias, regardless
+of the environment:
+
+    // modules/custom/Foo/foo.services.yml
+    services:
+      foo.storage:
+        class: 'MongoDB\Database'
+        factory: ['@mongodb.database_factory', 'get']
+        arguments: ['foodb']
+
+      foo.bar:
+        class: 'Drupal\foo\Bar'
+        arguments: ['@foo.storage', '$logger.channel.foo']
+
+This allows services in the module to access the database in both function code
+for Drupal hooks, and OO code for component-level logic without having to be
+environment-aware.
+
+If the `mongodb_watchdog` module is enabled, the logger passed to the
+application will be a standard PSR-3 logger writing to MongoDB without the
```

Comment:
Not really: the plain PSR-3 logger going only to MongoDB is "@mongodb.logger". The "normal" channel "@logger.channel.foo" is a Drupal logging channel like any other, which sends to MongoDB Watchdog, but also to any other enabled logger module. I think we need to make this clearer: for most usages, people will want to use the "normal" logging channel, as recommended by the Drupal API docs, not the dedicated MongoDB PSR-3 logger.